File: projekat/frontend/src/api/schedule.ts
import apiClient from "./client";

export type Frequency = "hourly" | "daily" | "weekly";

export type DriveFileStatus = "pending" | "importing" | "imported" | "skipped" | "failed";

export interface DriveFileProgress {
  name: string;
  status: DriveFileStatus;
}

export interface DriveSchedule {
  enabled: boolean;
  frequency: Frequency;
  hour: number;
  minute: number;
  day_of_week: number; // Mon=0 … Sun=6
  last_run_at: string | null;
  next_run_at: string | null;
  running: boolean;
  cancelling: boolean;
  last_result: string | null;
  files: DriveFileProgress[];
}

export type DriveScheduleUpdate = Pick<
  DriveSchedule,
  "enabled" | "frequency" | "hour" | "minute" | "day_of_week"
>;

export async function getDriveSchedule(): Promise<DriveSchedule> {
  const { data } = await apiClient.get<DriveSchedule>("/api/v1/schedule/drive");
  return data;
}

export async function updateDriveSchedule(
  body: DriveScheduleUpdate
): Promise<DriveSchedule> {
  const { data } = await apiClient.put<DriveSchedule>("/api/v1/schedule/drive", body);
  return data;
}

export async function cancelDriveImport(): Promise<DriveSchedule> {
  const { data } = await apiClient.post<DriveSchedule>("/api/v1/schedule/drive/cancel");
  return data;
}

File: projekat/backend/app/schemas/schedule.py
from __future__ import annotations

from datetime import datetime
from typing import Literal, Optional

from pydantic import BaseModel, ConfigDict, Field

Frequency = Literal["hourly", "daily", "weekly"]
Language = Literal["en", "bs", "de", "auto"]


class DriveFileProgress(BaseModel):
    name: str
    # pending | importing | imported | skipped | failed
    status: str


class DriveScheduleRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    enabled: bool
    frequency: Frequency
    hour: int
    minute: int
    day_of_week: int
    language: Language = "en"
    last_run_at: Optional[datetime] = None
    next_run_at: Optional[datetime] = None
    # Live runner state (not persisted) — lets the UI show a run in progress.
    running: bool = False
    cancelling: bool = False
    last_result: Optional[str] = None
    files: list[DriveFileProgress] = []


class DriveScheduleUpdate(BaseModel):
    enabled: bool
    frequency: Frequency
    hour: int = Field(ge=0, le=23)
    minute: int = Field(ge=0, le=59)
    day_of_week: int = Field(ge=0, le=6)
    language: Language = "en"
